examples/selective2d/2d_train.py

```diff
@@ -328,21 +331,30 @@ def pp_tp_train(stage, mesh, args):
     )
     local_iter_num = 0
     iter_time = 0.0
-    while local_iter_num < train_iters:
-        optimizer.zero_grad()
-        t0 = time.perf_counter()
-        X, Y = get_rand(args)
-        if pp_rank == 0:
-            out = stage(X)
-        elif pp_rank == args.pp_size - 1:
-            out = stage(Y)
-        else:
-            out = stage()
-        optimizer.step()
-        t1 = time.perf_counter()
-        dt = t1 - t0
-        local_iter_num += 1
-        iter_time += dt
+    with profile(
+        activities=[ProfilerActivity.CPU, ProfilerActivity.CUDA],
+        schedule=torch.profiler.schedule(
+            skip_first=5, wait=0, warmup=4, active=1, repeat=1
+        ),
+    ) as prof:
+        while local_iter_num < train_iters:
+            optimizer.zero_grad()
+            t0 = time.perf_counter()
+            X, Y = get_rand(args)
+            if pp_rank == 0:
+                out = stage(X)
+            elif pp_rank == args.pp_size - 1:
+                out = stage(Y)
```

Comment:
does this mean we have two pipeline here?

Reply:
we have args.pp_size stages but we only pass 'input' to the first stage and 'label' to the last stage (to calculate loss and backpropagation). intermediate values are automatically passed by pippy